### PR TITLE
[PD] Add custom memory pool option to support Mooncake PD with NVLink 

### PR DESCRIPTION
**Pull request by ShangmingCai**
## Motivation
This PR adds an env var option: SGLANG_MOONCAKE_CUSTOM_MEM_POOL, if it is set to true, mooncake will initialize the kvcache in a custom memory pool to enable cross-node NVLink transport.

**This feature requires mooncake-transfer-engine >= 0.3.3.post2**

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/disaggregation/mooncake/memory_pool.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+class CustomAllocator:
+    _instances: Dict[torch.device, CUDAPluggableAllocator] = {}
+    _lock: Final = threading.Lock()
+
+    @classmethod
+    def _get_so_path(cls) -> str:
+        """Dynamically locate hook.so in the mooncake package installation"""
+        try:
+            # Attempt to locate package resource
+            with resources.path("mooncake", "hook.so") as so_path:
```
Comment (fzyzcjy) [inline python/sglang/srt/disaggregation/mooncake/memory_pool.py:20]: nit: wondering whether the name needs to be more distinguishable (but anyway this is unrelated to SGLang side)

File: python/sglang/srt/disaggregation/mooncake/memory_pool.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import os
+import threading
+from importlib import resources
+from typing import Dict, Final, Optional
+
+import torch
+from torch.cuda.memory import CUDAPluggableAllocator
+
+
+class CustomAllocator:
```
Comment (fzyzcjy) [inline python/sglang/srt/disaggregation/mooncake/memory_pool.py]: nit: wondering whether this should be done on mooncake side or sglang side. Suppose one day another user of mooncake wants to do sth similar on their kv cache, then seems they will directly reuse this code

File: python/sglang/srt/disaggregation/utils.py [hunk 3]
```diff
@@ -84,24 +84,59 @@ def free(self, free_index: int):
...
+                self.output_token_logprobs_idx = torch.zeros(
+                    (size, 16), dtype=torch.int32, device="cuda"
+                )
+                self.output_top_logprobs_val = torch.zeros(
+                    (size, max_top_logprobs_num), dtype=torch.float32, device="cuda"
+                )
+                self.output_top_logprobs_idx = torch.zeros(
+                    (size, max_top_logprobs_num), dtype=torch.int32, device="cuda"
+                )
+        else:
```
Comment (fzyzcjy) [inline python/sglang/srt/disaggregation/utils.py]: nit: maybe avoid code copy-paste, e.g.

```
with (torch.cuda.use_mem_pool(self.custom_mem_pool) if enable the custom pool else nullcontext():
   ... old code ...
```

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 4]
```diff
@@ -277,22 +291,42 @@ def _create_buffers(self):
...
-                )
-                for _ in range(self.layer_num)
-            ]
+            if self.enable_custom_mem_pool:
+                self.k_buffer = []
+                self.v_buffer = []
+
+                assert self.custom_mem_pool is not None
+                with torch.cuda.use_mem_pool(self.custom_mem_pool):
+                    for _ in range(self.layer_num):
```
Comment (fzyzcjy) [inline python/sglang/srt/mem_cache/memory_pool.py]: same, would be great to avoid code dup

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 5]
```diff
@@ -569,16 +606,41 @@ def __init__(
...
-                    dtype=self.store_dtype,
-                    device=device,
-                )
-                for _ in range(layer_num)
-            ]
+            if self.enable_custom_mem_pool:
+                self.kv_buffer = []
+
+                assert self.custom_mem_pool is not None
+                with torch.cuda.use_mem_pool(self.custom_mem_pool):
```
Comment (fzyzcjy) [inline python/sglang/srt/mem_cache/memory_pool.py]: same

**Comment by fzyzcjy** [comment]
Only some nits, and ready to merge after some very quick cleanup

File: python/sglang/srt/disaggregation/mooncake/memory_pool.py (hunk 1 above)
Comment (ShangmingCai) [reply python/sglang/srt/disaggregation/mooncake/memory_pool.py:20]: The name issue will be optimized in the future release, since v0.3.3.post2 has been released. Will fix this before v0.3.4, and make a PR to fix this at that time.

File: python/sglang/srt/disaggregation/mooncake/memory_pool.py (hunk 2 above)
Comment (ShangmingCai) [reply python/sglang/srt/disaggregation/mooncake/memory_pool.py]: ditto. Will fix this in v0.3.4.

File: python/sglang/srt/disaggregation/utils.py [hunk 6]
```diff
@@ -84,24 +85,54 @@ def free(self, free_index: int):
...
+            else nullcontext()
+        ):
+            # TODO: abort top_logprobs_num > 128 in PD
+
+            # We transfer the metadata of first output token to decode
+            # The minimal size for RDMA is 64Bytes, so we pad it to > 64Bytes
+            self.output_ids = torch.zeros(
+                (size, 16),
+                dtype=torch.int32,
+                device="cuda" if self.custom_mem_pool else "cpu",
```
Comment (fzyzcjy) [inline python/sglang/srt/disaggregation/utils.py]: nit: what about extracting a variable `"cuda" if self.custom_mem_pool else "cpu",` and then do `device=device` everywhere

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 7]
```diff
@@ -260,6 +262,23 @@ def __init__(
 
         self.head_num = head_num
         self.head_dim = head_dim
+
+        # for disagg with nvlink
+        is_custom_mem_pool_enabled = os.environ.get("SGLANG_MOONCAKE_CUSTOM_MEM_POOL")
+        self.enable_custom_mem_pool = (
+            is_custom_mem_pool_enabled is not None
+            and is_custom_mem_pool_enabled.lower() == "true"
+        )
```
Comment (fzyzcjy) [inline python/sglang/srt/mem_cache/memory_pool.py]: get_bool_env_var

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 8]
```diff
@@ -277,22 +296,27 @@ def _create_buffers(self):
...
+                        dtype=self.store_dtype,
+                        device=self.device,
+                    )
+                    v = torch.zeros(
+                        (self.size + self.page_size, self.head_num, self.head_dim),
+                        dtype=self.store_dtype,
+                        device=self.device,
+                    )
+                    self.k_buffer.append(k)
+                    self.v_buffer.append(v)
```
Comment (fzyzcjy) [inline python/sglang/srt/mem_cache/memory_pool.py]: maybe just keep old code, and the only thing that changes is a `with` (and some indents)

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 9]
```diff
@@ -569,16 +596,38 @@ def __init__(
...
             # The padded slot 0 is used for writing dummy outputs from padded tokens.
-            self.kv_buffer = [
-                torch.zeros(
-                    (size + page_size, 1, kv_lora_rank + qk_rope_head_dim),
-                    dtype=self.store_dtype,
-                    device=device,
-                )
-                for _ in range(layer_num)
-            ]
+            self.kv_buffer = []
```
Comment (fzyzcjy) [inline python/sglang/srt/mem_cache/memory_pool.py]: same

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 10]
```diff
@@ -569,16 +596,38 @@ def __init__(
         self.kv_lora_rank = kv_lora_rank
         self.qk_rope_head_dim = qk_rope_head_dim
 
+        # for disagg with nvlink
+        is_custom_mem_pool_enabled = os.environ.get("SGLANG_MOONCAKE_CUSTOM_MEM_POOL")
```
Comment (fzyzcjy) [inline python/sglang/srt/mem_cache/memory_pool.py]: same

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 11]
```diff
@@ -569,16 +596,38 @@ def __init__(
...
+            is_custom_mem_pool_enabled is not None
+            and is_custom_mem_pool_enabled.lower() == "true"
+        )
+        if self.enable_custom_mem_pool:
+            from sglang.srt.disaggregation.mooncake.memory_pool import (
+                MooncakeNVLinkAllocator,
+            )
+
+            allocator = MooncakeNVLinkAllocator.get_allocator(self.device)
+            self.custom_mem_pool = torch.cuda.MemPool(allocator.allocator())
```
Comment (fzyzcjy) [inline python/sglang/srt/mem_cache/memory_pool.py:609]: shall we put such common logic to base class

**Review by fzyzcjy — COMMENTED** [review]
some minor nits

File: python/sglang/srt/mem_cache/memory_pool.py (hunk 11 above)
Comment (ShangmingCai) [reply python/sglang/srt/mem_cache/memory_pool.py:609]: Sure, I think @ByronHsu will need this too. Maybe it would be better to implement this in another PR after I talk with him, so that it will be less painful for him to sync the code.

File: python/sglang/srt/disaggregation/mooncake/memory_pool.py (hunk 1 above)
Comment (fzyzcjy) [reply python/sglang/srt/disaggregation/mooncake/memory_pool.py:20]: nit: then maybe add a TODO comment to MooncakeNVLinkAllocator class

File: python/sglang/srt/mem_cache/memory_pool.py (hunk 11 above)
Comment (fzyzcjy) [reply python/sglang/srt/mem_cache/memory_pool.py:609]: ok looks reasonable

**Review by fzyzcjy — APPROVED** [review]

**Comment by fzyzcjy** [comment]
LGTM, the nits are optional
